File: android/companion-core/src/main/kotlin/com/xopmc/galaxybridge/core/PairingTranscript.kt
package com.xopmc.galaxybridge.core

import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.security.spec.ECParameterSpec

object P256Keys {
    const val X963_LENGTH = 65

    fun x963Representation(publicKey: PublicKey): ByteArray {
        val ecKey = publicKey as? ECPublicKey ?: throw IllegalArgumentException("not an EC public key")
        return byteArrayOf(0x04) + fixedUnsigned(ecKey.w.affineX) + fixedUnsigned(ecKey.w.affineY)
    }

    fun publicKey(x963: ByteArray): PublicKey {
        require(x963.size == X963_LENGTH && x963[0] == 0x04.toByte()) { "invalid P-256 X9.63 key" }
        val parameters = AlgorithmParameters.getInstance("EC").apply {
            init(ECGenParameterSpec("secp256r1"))
        }.getParameterSpec(ECParameterSpec::class.java)
        val point = ECPoint(
            BigInteger(1, x963.copyOfRange(1, 33)),
            BigInteger(1, x963.copyOfRange(33, 65)),
        )
        return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(point, parameters))
    }

    private fun fixedUnsigned(value: BigInteger): ByteArray {
        val encoded = value.toByteArray()
        val unsigned = if (encoded.size == 33 && encoded[0] == 0.toByte()) encoded.copyOfRange(1, 33) else encoded
        require(unsigned.size <= 32)
        return ByteArray(32 - unsigned.size) + unsigned
    }
}

object PairingTranscript {
    private val domain = "GalaxyBridge/Pairing/v1".encodeToByteArray()
    private val requestDomain = "GalaxyBridge/PairingRequest/v1".encodeToByteArray()
    private val commitDomain = "GalaxyBridge/PairingCommit/v1".encodeToByteArray()
    private val commitAcknowledgementDomain = "GalaxyBridge/PairingCommitAck/v1".encodeToByteArray()

    fun makeRequest(
        token: ByteArray,
        clientNonce: ByteArray,
        macPublicKey: ByteArray,
        displayName: ByteArray,
    ): ByteArray = encode(listOf(requestDomain, token, clientNonce, macPublicKey, displayName))

    fun make(
        token: ByteArray,
        clientNonce: ByteArray,
        serverNonce: ByteArray,
        macPublicKey: ByteArray,
        androidPublicKey: ByteArray,
    ): ByteArray {
        return encode(listOf(domain, token, clientNonce, serverNonce, macPublicKey, androidPublicKey))
    }

    fun makeCommit(
        token: ByteArray,
        clientNonce: ByteArray,
        serverNonce: ByteArray,
        macPublicKey: ByteArray,
        androidPublicKey: ByteArray,
        hostId: String,
        deviceId: String,
        sessionId: String,
    ): ByteArray = encode(
        listOf(
            commitDomain,
            token,
            clientNonce,
            serverNonce,
            macPublicKey,
            androidPublicKey,
            hostId.encodeToByteArray(),
            deviceId.encodeToByteArray(),
            sessionId.encodeToByteArray(),
        ),
    )

    fun makeCommitAcknowledgement(
        commitTranscript: ByteArray,
        commitSignature: ByteArray,
    ): ByteArray = encode(listOf(commitAcknowledgementDomain, commitTranscript, commitSignature))

    private fun encode(fields: List<ByteArray>): ByteArray {
        val output = ByteArrayOutputStream()
        DataOutputStream(output).use { writer ->
            fields.forEach { field ->
                writer.writeInt(field.size)
                writer.write(field)
            }
        }
        return output.toByteArray()
    }

    fun verify(signatureDer: ByteArray, transcript: ByteArray, publicKeyX963: ByteArray): Boolean =
        runCatching {
            Signature.getInstance("SHA256withECDSA").run {
                initVerify(P256Keys.publicKey(publicKeyX963))
                update(transcript)
                verify(signatureDer)
            }
        }.getOrDefault(false)
}

object SessionAuthenticationTranscript {
    private val domain = "GalaxyBridge/SessionAuthentication/v1".encodeToByteArray()

    fun make(
        deviceId: String,
        sessionId: String,
        nonce: ByteArray,
        timestampUnixSeconds: Long,
        identityPublicKey: ByteArray,
    ): ByteArray {
        val timestamp = ByteArrayOutputStream().also { output ->
            DataOutputStream(output).use { it.writeLong(timestampUnixSeconds) }
        }.toByteArray()
        val output = ByteArrayOutputStream()
        DataOutputStream(output).use { writer ->
            listOf(
                domain,
                deviceId.encodeToByteArray(),
                sessionId.encodeToByteArray(),
                nonce,
                timestamp,
                identityPublicKey,
            ).forEach { field ->
                writer.writeInt(field.size)
                writer.write(field)
            }
        }
        return output.toByteArray()
    }
}

File: Sources/GalaxyBridgeCore/PairingTranscript.swift
import CryptoKit
import Foundation

public enum PairingTranscript {
    private static let domain = Data("GalaxyBridge/Pairing/v1".utf8)
    private static let requestDomain = Data("GalaxyBridge/PairingRequest/v1".utf8)
    private static let commitDomain = Data("GalaxyBridge/PairingCommit/v1".utf8)
    private static let commitAcknowledgementDomain = Data("GalaxyBridge/PairingCommitAck/v1".utf8)

    public static func makeRequest(
        token: Data,
        clientNonce: Data,
        macPublicKey: Data,
        displayName: Data
    ) -> Data {
        encode([requestDomain, token, clientNonce, macPublicKey, displayName])
    }

    public static func make(
        token: Data,
        clientNonce: Data,
        serverNonce: Data,
        macPublicKey: Data,
        androidPublicKey: Data
    ) -> Data {
        encode([domain, token, clientNonce, serverNonce, macPublicKey, androidPublicKey])
    }

    public static func makeCommit(
        token: Data,
        clientNonce: Data,
        serverNonce: Data,
        macPublicKey: Data,
        androidPublicKey: Data,
        hostID: String,
        deviceID: String,
        sessionID: String
    ) -> Data {
        encode([
            commitDomain,
            token,
            clientNonce,
            serverNonce,
            macPublicKey,
            androidPublicKey,
            Data(hostID.utf8),
            Data(deviceID.utf8),
            Data(sessionID.utf8),
        ])
    }

    public static func makeCommitAcknowledgement(
        commitTranscript: Data,
        commitSignature: Data
    ) -> Data {
        encode([commitAcknowledgementDomain, commitTranscript, commitSignature])
    }

    private static func encode(_ fields: [Data]) -> Data {
        var result = Data()
        for field in fields {
            precondition(field.count <= Int(UInt32.max))
            var length = UInt32(field.count).bigEndian
            withUnsafeBytes(of: &length) { result.append(contentsOf: $0) }
            result.append(field)
        }
        return result
    }

    public static func verify(
        signatureDER: Data,
        transcript: Data,
        publicKeyX963: Data
    ) throws -> Bool {
        let publicKey = try P256.Signing.PublicKey(x963Representation: publicKeyX963)
        let signature = try P256.Signing.ECDSASignature(derRepresentation: signatureDER)
        return publicKey.isValidSignature(signature, for: transcript)
    }
}
